### DISCUSSION TITLE
FACS RnD

### DISCUSSION THREAD

**Discussion by benngarnish** [jjburton/cgmTools#162] [General]
Thought it would be good to start having solid conversations here. 
Latest issue post: https://github.com/jjburton/cgmTools/issues/159#issuecomment-744751064

At some point today I’ll be looking at naming joints for the cage. For this I’ll be referencing muscles on the face and a lot of the hard work that already gone into FACS from various people in the interwebz

**Comment by jjburton** [comment]
### Cage | 12.15.2020
Picking up @benngarnish 's pass. Building cage...

- Adding a few points I think we need on the cage
- Going to work on the simple placement step

![image](https://user-images.githubusercontent.com/48218358/102266671-60a4b180-3ede-11eb-9d45-53cef2f886aa.png)

This is the pass one define plan.
![image](https://user-images.githubusercontent.com/48218358/102290407-9cec0800-3f06-11eb-94f3-1dcfbea2e267.png)

![image](https://user-images.githubusercontent.com/48218358/102290385-8cd42880-3f06-11eb-8100-f048e9c8b2d7.png)

### Poses
- Setup a tmp file for team to start messing with poses. 
- I'd like to get a simple set to start testing with, maybe the lowest tier set (we need those set lists too :)
- Wanted to get you all something you could start iterating with
-
mrsMakers_facs\content\RD\poses\scenes\pass_01\poses_pass_01_01.mb
![image](https://user-images.githubusercontent.com/48218358/102308472-4f829180-3f2c-11eb-8031-4e2670d0a308.png)

**Comment by jjburton** [comment]
### Define State

- [x] Define Anchors
- [x] Sub handles

- Try having driver loc in scale space for overall shaping
    - This works well 
- distance constrain vs curve attach

### Pass 1
https://mobile.twitter.com/CgMonastery/status/1339301930280103936

- Constraining handles works well.
- Too dense for pass one still. 

for next...

- [x] Need main anchors
- [x] Eye sizers
- [x] Simplify main movers

![image](https://user-images.githubusercontent.com/48218358/102402171-0113d800-3faa-11eb-9fde-33deae0cddd9.png)

### Pass 2
- Got eye sphere in from eye block -- factored to Blockshapes
- The locator forms not working well visually with where we're going
- Going to add a new lolihandle like pre handles....that maybe confusing at prerig.... goint to try offset semispheres..

![image](https://user-images.githubusercontent.com/48218358/102444715-7950bc80-3fef-11eb-8336-b41d12b307b5.png)

```python
#...snip working through eyeOrb call
reload(BLOCKSHAPES)
BLOCKSHAPES.eyeOrb(mBlock,mCtrl,mStateNull) 
mCtrl = cgmMeta.asMeta('L_eyeAnchor_defineHandle')
mStateNull = mBlock.noTransDefineNull
```
![image](https://user-images.githubusercontent.com/48218358/102447267-4dd0d080-3ff5-11eb-9516-7e82168c6399.png)
- Switched to circle anchors
- Next I'll wire the initial pass of handles to these so they'll follow these around
- Then I'll add a pass to build the second layer of handles

Pass 2 result
https://mobile.twitter.com/CgMonastery/status/1339663037180112897
RD/baseMale/template/pass_01/baseMale_pass_01_03.mb
Much better for the most part.

- [x] Need more trace lines. Around full edge for one
- [ ] Going to need a few more main handles to better frame things. 5 for the eyelines to be more exact and make form state less tedious. With. 
- [x] Nose top main handle
- [x] Brow peaks
- [ ] Eye innerMid
- [ ] Eye Outermid
- [x] noseTip (for curvature)
- [ ] Move the handle builds to a sep call and not the base
- [ ] Add snap
- [ ] Need lip inners

```python
#...for grabbing tag lists for curve creation
l = []
for mObj in cgmMeta.asMeta(sl=1):
    l.append(str(mObj.handleTag))
print l
```

### Pass 3
Working through stuff. 
![image](https://user-images.githubusercontent.com/48218358/102580673-a02cf280-40c4-11eb-85c2-bc1b769ce323.png)

- [x] Iris/pupil
- [x] Iris/pupil depth attrs
- [x] eye handle
- [x] wire eye and iris/pupil for mirror at define

- [x] 2x jointSize for handles
- [x] No scale on anchors
- [ ] Lose anchor naming if we aren't going to use other for deform?
- [x] Wire those
- [x] Pull sub handles from define
- [x] Anchors - nostril, cheekFront (better name), orbital, cheek
- [ ] Add teeth anchor
- [ ] Add mouth back

Thinking the more detailed points will be the form state and you must generate a mesh before moving to prerig. So that workflow generally would be
- Template | size eyes, position anchor points
- Form | Builds sub handles for shaping
- Form | Must build mesh before you go to prerig. It'll be a right click menu item

Define state decent. Will hit form this weekend which will be the fuller form as before.
https://twitter.com/CgMonastery/status/1340174117597835265

**Reply by jjburton** [reply]
@benngarnish | I'm thinking perhaps just skinning the cage to my shaper handles might work for moving it around. I'm gonna try that before going into the generating a point of data for every vert rabbit hole 👍

**Reply by benngarnish** [reply]
Let me know how this goes, but sounds interesting!

**Comment by benngarnish** [comment]
This is looking great. 
I have some options for controls:

1.
![B568298D-5C71-4BCB-BDCD-45CA5B3EAB36](https://user-images.githubusercontent.com/1122587/102411932-aea6db80-3fea-11eb-876b-655211b0a6ed.jpeg)

2.
![E8EA176C-4540-4132-9E45-6C16D037BA2C](https://user-images.githubusercontent.com/1122587/102411964-be262480-3fea-11eb-9c18-82ecbaa6fc0c.jpeg)

- this would require some attributes in the channel box that you would dial up and down.

**Reply by jjburton** [reply]
Interesting, I'll have to play with the setups and see. I'll put a call out in the next weekly update on monday asking for help. Hopefully we can get through define to form state by then as well.

**Reply by SplineO** [reply]
I'm using the second option on my rigs .. more or less, with major controls using the Osipa method. Having all those attributes is something most of our animators find unintuitive, though. I always wanted to try Option 1, but never had the chance! ;)

**Reply by benngarnish** [reply]
Thanks for the feedback. It’s very useful.

**Reply by jjburton** [reply]
@benngarnish | can you...

- [ ] Setup a clean file with the blendshapes all wired to the head and a poseBuffer. RD.poseBuffer.cage_01, etc
- [ ] Ref in that pose buffer head with your control setup and wire to the head? RD.controls.benn, etc

This will make it easier to evaluate and for others to connect different setups.

Thoughts?

**Reply by benngarnish** [reply]
Love it. That’s my thought!
This will make it easier for existing setups to be ported across to mrs

**Reply by jjburton** [reply]
We might want to have a couple of different blend sets - realistic/ toony  to start depending on how the shapes look on some of the other heads.

**Reply by benngarnish** [reply]
I think that people should have the ability to create a base from which they can start with. 
so if someone has a really cartoony look they can do one face and then save this out as a base for future rigs?

**Reply by jjburton** [reply]
Of course. I'm just talking about the initials blendsets. It might be good to have some that are more extreme. No worries, we can work through when we get there.

**Comment by benngarnish** [comment]
![EE6FFF8D-8030-4A0B-80E2-8167ACBBD050](https://user-images.githubusercontent.com/1122587/102540407-e4a99580-40a6-11eb-97b8-7e812eff67a3.jpeg)

inspired

**Reply by jjburton** [reply]
That looks really cool. Something that complicated might work decently with a pose library.

**Comment by jjburton** [comment]
### Define revisit
![image](https://user-images.githubusercontent.com/48218358/102698583-1dab4c80-4204-11eb-9f62-3f09bc4536fe.png)

- [x] Need a few more base points and
- [x] Trace lines really help understand what's what

![image](https://user-images.githubusercontent.com/48218358/102704580-e9a34c00-4242-11eb-940e-40f0d86461fe.png)
- might add the under neck or swap that with the bottom one

### Form
Plan going in
![image](https://user-images.githubusercontent.com/48218358/102694727-8c7aac80-41e8-11eb-94e6-111c1704457b.png)

- [ ] Form handles structure
- [ ] Add new handles

Form To do:

- [x] Define curves template on form state
- [x] Define curves template off on form delete
- [x] Gather define handles for remapping -- as done before
- [x] Call to collect define state pos dat (relative dat to work through that?)
- [x] Call to restore
- [x] Build stuff by sections
- [x] Brow
- [x] Nose
- [x] Eyes
- [ ] Mouth
- [ ] Other
- [ ] Test skinning cage to label joints on handles

```python
reload(FACSUTILS)
FACSUTILS.get_tagList()#...reg tag list
FACSUTILS.get_tagList(True)#...repeate front to loop

FACSUTILS.get_driverTags()#...for constraints
FACSUTILS.iterNameTags('noseShape','R')#...iter with side
FACSUTILS.iterNameTags('mouthTop')#...iter tag
FACSUTILS.iterNameTags('mouthBottom')

#...getting lide and center tags from a custom dict
l = []
l_center = []
for k in FACSUTILS.d_formScaleSpace['default']:
    if k.startswith('R_'):
        l.append(k.replace('R_',''))
    else:
        l_center.append(k)
l.sort()
l_center.sort()
print( l)
print l_center


reload(BLOCKSHAPES)
BLOCKSHAPES.eyeOrb(mBlock,mCtrl,mStateNull) 
mCtrl = cgmMeta.asMeta('L_eyeAnchor_defineHandle')
mStateNull = mBlock.noTransDefineNull

#...finding bad tags
ml_new = cgmMeta.asMeta(sl=1)
ml_good = []
for mObj in ml_new:
    try:
        if mObj.mNode:
            ml_good.append(mObj)
    except:pass

l_tags = []
l_dups = []
d_tags = {}
for mObj in ml_good:
    try:mObj
    except:break
    if mObj.handleTag not in l_tags:
        l_tags.append(mObj.handleTag)
        d_tags[mObj.handleTag] = mObj
    else:
        print mObj
        print d_tags[mObj.handleTag]
        l_dups.extend([mObj,d_tags[mObj.handleTag]])
mc.select([mObj.mNode for mObj in l_dups])
mc.select([mObj.mNode for mObj in ml_good])
```

### Define pass again
- added jaw front, mouth interior, nose tip, lid mains
- at define state added visual differentiation between the main and constrained anchors - main are the squircles and the constrained the circles. used open cubes for mouth interior stuff
![image](https://user-images.githubusercontent.com/48218358/102853844-bd283500-43e6-11eb-8379-5081c3afb9c8.png)

### Form progress 12.21
- All the new makers are scale spaced and creating
- working through the constraint setup, hence the reworking some define anchors to have more data there. 
- Decent progress but this is pretty dense work and is going slower than I'd hoped. Kids are home for the holidays so that makes it harder :)

**Comment by jjburton** [comment]
### Form work 12.22
https://twitter.com/CgMonastery/status/1341416429158064134
- handles all mapped to anchors.


Need to add a few form handles after a pass on curves

- [x] skull front
- [x] sneerLower
- [x] noseTopCenter
- [x] L_lipUprBow
- [ ] visForMouthInterior stuff to simplify

Quick skin to the form labels...it's working :)
![image](https://user-images.githubusercontent.com/48218358/102935583-10909680-446c-11eb-9d07-36d100ff496a.png)

There seems to be some cap with the number of attributes a node can have wired and as slow down that happens when it's selected as maya catches up. I'm not sure if this is controller wiring or a number of connections. Noticed this on my facs block as the number of connections is large.

As such, I'm trying something a little different. Rather that have all the main connections on the single node, I'm going to move some to the statenulls and see where we get.

Additionally I'm testing not using msgLists for places that don't need them. They're fine generally but I'm thinking having 200+ message attributes wired on a given node might have been part of it. So....

- [x] Define | change handles and curves to block.defineNull.curves/handles wiring
- [x] Make self mirror calls work here
- [x] Make sure state checks work with the new setup

Other bits
- for loop deletions are slow, just [mObj.mNode for mObj in ml]

### 12.23
With the new wiring for handles, the blockDat isn't working. need to patch this until new config setup done.

Forming to morphy kid and roller girl to test...
- [x] BLOCKUTILS.controls_get | account for new setup
- [x] Make the rotation of define handles not muck up the form going into form (bake in those rotations)
- [x] Add frame for utils to block editor
- [x] Need some more anchors
- [x] gum upr/lwr
- [x] cheekSide
- [x] upr/lwr peak
- [x] teeth back

Added frame work for module utils in a block module
![image](https://user-images.githubusercontent.com/48218358/103061250-5fbfef80-4570-11eb-8612-d234adcd9ab9.png)

**Comment by jjburton** [comment]
### 01.05.2021

- Benn generated poses for Josh
- Josh needs to move to pre and try working in posed head.

**Comment by jjburton** [comment]
### 01.11.2021
Aight, what do we need to do here.

### Joysticks
- [ ] Need to write a hook up setup to test wiring
- [ ] Figure out our base set

### FACS Block
- [x]  bring in Benn's bs'd head and get that connecting to the form state
- [ ] evaluate this functionality 
- [ ] Push to prerig to setup our joints
- [ ] Skeleton state

### Other RD
- [ ] rebuild Drake's head with sdk groups
- [ ] Try using shaped bs head to drive drakes head
- [ ] Setup Drake joystick setup

### Calls needed
- [ ] Controller creation
- [ ] Wiring framework


### Thinking
At form state to attach the geo we'd need to :
-enforce a baseSize, scale and positioning to match the base male
- reset and store pose call to prep
- bind call
- bring in head geo with blends
- snap to the rigBlock
- don't need poseBuffer at this point, that might be in the way here
- reposition the handles to the original places

###Research
- Buddy sent an animal link which is pretty sweet - https://www.animalfacs.com/
